### PR TITLE
Fix multithread writing for fragment result cache

### PR DESCRIPTION
**Pull request by superqtqt**
== RELEASE NOTES ==

General Changes
* Fix multithread writing for fragment result cache  #16455 
* add config `fragment-result-cache.max-single-pages-size` to control the max fragement cache file size

### CODE REVIEW

**Comment by superqtqt** [comment]
@shixuan-fan @kaikalur @highker @ajaygeorge @tdcmeehan 
can you help me review the code

**Comment by superqtqt** [comment]
@shixuan-fan can you help me merge the code

File: presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java [hunk 1]
```diff
@@ -146,13 +152,18 @@ private static long getPagesSize(List<Page> pages)
...
-    private void cachePages(CacheKey key, Path path, List<Page> pages)
+    private void cachePages(CacheKey key, Path path, List<Page> pages, long resultSize)
     {
         // TODO: To support both memory and disk limit, we should check cache size before putting to cache and use written bytes as weight for cache
         try {
             Files.createFile(path);
             try (SliceOutput output = new OutputStreamSliceOutput(newOutputStream(path, APPEND))) {
-                writePages(pagesSerde, output, pages.iterator());
+                if (isBlockEncodingCompressionEnabled) {
+                    writePages(pagesSerdeFactory.createPagesSerde(), output, pages.iterator());
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java:162]: Curious: Does it make sense that instead of creating `PagesSerde` every time, we create only once in the constructor, and directly replace current `pagesSerde`

**Review by shixuan-fan — CHANGES_REQUESTED** [review]
Thanks for fixing this. Overall looks good just some nits.

File: presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java (hunk 1 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java:162]: Alternatively, you could use something similar to the changes I made in Trino https://github.com/trinodb/trino/pull/5545 which (mostly) puts the not thread safe buffers into a context object that can be created / released. I say "mostly" here because I didn't move the `Compressor` instance in which you would need to do in order to use it that way.

Another thing worth pointing out is that this implementation, as written, doesn't support encryption of the serialized pages.

File: presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java [hunk 2]
```diff
@@ -67,6 +67,9 @@
 
     private final Path baseDirectory;
     private final long maxInFlightBytes;
+    private final long maxSinglePagesBytes;
+    private final Boolean isBlockEncodingCompressionEnabled;
```
[the remark this answers is not in the document]
Comment (superqtqt) [reply presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java]: thank you;fixed

File: presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java (hunk 1 above)
Comment (superqtqt) [reply presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java]: @shixuan-fan i have delete the write compress buffer and relation code.I find it not work;

@pettyjamesm thanks for you advice .

File: presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java [hunk 3]
```diff
@@ -199,7 +203,8 @@ private static void tryDeleteFile(Path path)
             fragmentCacheStats.incrementCacheHit();
             return Optional.of(closeWhenExhausted(result, inputStream));
         }
-        catch (UncheckedIOException | IOException e) {
+        catch (Throwable e) {
+            log.error(e, "read path %s error", path);
```
Comment (pettyjamesm) [inline presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java:207]: You don’t want to catch Throwable here- since “should be fatal” members of the Error subheirarchy like OutOfMemoryError are not what you want to catch here. Seems like catching Exception is sufficient here instead.

Comment (superqtqt) [reply presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java:207]: Yes,I lack consideration.I hava fixed

File: presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java (hunk 1 above)
Comment (shixuan-fan) [reply presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java]: > @shixuan-fan i have delete the write compress buffer and relation code.I find it not work.Can you help me approve

I'm trying to understand this. Do you mean you tried @pettyjamesm 's suggestion but it does not work?

File: presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java (hunk 3 above)
Comment (shixuan-fan) [reply presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java:207]: Should we consider only extending it to `RuntimeException`? I don't think we should cover checked exception here.

Also, does it make sense to catch this separately from `IOException` and `UncheckedIOException`? There are legit reasons for them to happen, but maybe not for other exceptions?

File: presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java [hunk 4]
```diff
@@ -88,7 +90,9 @@ public FileFragmentResultCacheManager(
 
         this.baseDirectory = Paths.get(cacheConfig.getBaseDirectory());
         this.maxInFlightBytes = cacheConfig.getMaxInFlightSize().toBytes();
-        this.pagesSerde = new PagesSerdeFactory(blockEncodingSerde, cacheConfig.isBlockEncodingCompressionEnabled()).createPagesSerde();
+        this.maxSinglePagesBytes = cacheConfig.getMaxSinglePagesSize().toBytes();
+        this.pagesSerdeFactory = new PagesSerdeFactory(blockEncodingSerde, cacheConfig.isBlockEncodingCompressionEnabled());
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java:94]: Is this change still needed? Maybe it would be sufficient to keep `pagesSerde` only?

**Comment by shixuan-fan** [comment]
Also, it might be great to add a test in `TestFileFragmentResultCacheManager`

File: presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java (hunk 4 above)
Comment (superqtqt) [reply presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java:94]: the `pagesSerde` is not thread safe ,when writing data use a share buffer,and read is noe use the share buffer

File: presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java (hunk 3 above)
Comment (superqtqt) [reply presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java:207]: thank you.

File: presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java [hunk 5]
```diff
@@ -133,18 +157,40 @@ public void testBasic()
...
-                assertTrue(page1.getBlock(i).equals(0, 0, page2.getBlock(i), 0, 0, page1.getBlock(0).getSliceLength(0)));
-            }
+        ImmutableList.Builder<Future<String>> futures = ImmutableList.builder();
+        int i = 0;
+        int threadCnt = 10;
+        while (i < threadCnt) {
+            Future<String> future = writeExecutor.submit(() -> {
+                try {
+                    System.out.println(Thread.currentThread().getName());
+                    String threadInfo = String.format(WRITE_THREAD_NAME_FORMAT, Thread.currentThread().getName(), Thread.currentThread().getId());
```
Comment (shixuan-fan) [inline presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java:170]: Curious: is `%s,%s` enough?

**Review by shixuan-fan — COMMENTED** [review]
Some nits on test. Mostly look good!

File: presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java [hunk 6]
```diff
@@ -133,18 +157,40 @@ public void testBasic()
...
-        while (pages1.hasNext() && pages2.hasNext()) {
-            Page page1 = pages1.next();
-            Page page2 = pages2.next();
-            assertEquals(page1.getChannelCount(), page2.getChannelCount());
-            for (int i = 0; i < page1.getChannelCount(); i++) {
-                assertTrue(page1.getBlock(i).equals(0, 0, page2.getBlock(i), 0, 0, page1.getBlock(0).getSliceLength(0)));
-            }
+        ImmutableList.Builder<Future<String>> futures = ImmutableList.builder();
+        int i = 0;
+        int threadCnt = 10;
```
[the remark this answers is not in the document]
Comment (superqtqt) [reply presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java]: i hava add comment
and delete threadCnt

File: presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java [hunk 7]
```diff
@@ -133,18 +157,40 @@ public void testBasic()
...
+        while (i < threadCnt) {
+            Future<String> future = writeExecutor.submit(() -> {
+                try {
+                    System.out.println(Thread.currentThread().getName());
```
[the remark this answers is not in the document]
Comment (superqtqt) [reply presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java]: sorry ,i hava deleted

File: presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java [hunk 8]
```diff
@@ -133,18 +157,40 @@ public void testBasic()
...
+                    String threadInfo = String.format(WRITE_THREAD_NAME_FORMAT, Thread.currentThread().getName(), Thread.currentThread().getId());
+                    List<Page> pages = ImmutableList.of(new Page(createStringsBlock(threadInfo)));
+                    threadWriteCacheManager.put(threadInfo, SPLIT_2, pages).get();
+                    Optional<Iterator<Page>> result = threadWriteCacheManager.get(threadInfo, SPLIT_2);
+                    assertTrue(result.isPresent());
+                    assertPagesEqual(result.get(), pages.iterator());
+                    return "success";
```
[the remark this answers is not in the document]
Comment (superqtqt) [reply presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java]: ok finish

File: presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java [hunk 9]
```diff
@@ -133,18 +157,40 @@ public void testBasic()
...
+                catch (Exception e) {
+                    return null;
+                }
+            });
+            futures.add(future);
+            i++;
         }
-        assertFalse(pages1.hasNext());
-        assertFalse(pages2.hasNext());
+        assertEquals(futures.build().stream().filter(future -> {
```
[the remark this answers is not in the document]
Comment (superqtqt) [reply presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java]: ok finish

File: presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java [hunk 10]
```diff
@@ -51,19 +53,41 @@
...
+    {
+    }, new TestingSplit(2));
 
     private final ExecutorService writeExecutor = newScheduledThreadPool(5, daemonThreadsNamed("test-cache-flusher-%s"));
     private final ExecutorService removalExecutor = newScheduledThreadPool(5, daemonThreadsNamed("test-cache-remover-%s"));
 
     private URI cacheDirectory;
+    private FileFragmentResultCacheManager threadWriteCacheManager;
+
+    private static void assertPagesEqual(Iterator<Page> pages1, Iterator<Page> pages2)
```
[the remark this answers is not in the document]
Comment (superqtqt) [reply presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java]: fix,thank you

File: presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java [hunk 11]
```diff
@@ -51,19 +53,41 @@
...
+        assertFalse(pages2.hasNext());
+    }
 
     @BeforeClass
     public void setup()
             throws Exception
     {
         this.cacheDirectory = createTempDirectory("cache").toUri();
+        FragmentCacheStats stats = new FragmentCacheStats();
+        this.threadWriteCacheManager = fileFragmentResultCacheManager(stats);
```
[the remark this answers is not in the document]
Comment (superqtqt) [reply presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java]: fix,thank you

File: presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java [hunk 12]
```diff
@@ -133,18 +157,40 @@ public void testBasic()
...
-            Page page1 = pages1.next();
-            Page page2 = pages2.next();
-            assertEquals(page1.getChannelCount(), page2.getChannelCount());
-            for (int i = 0; i < page1.getChannelCount(); i++) {
-                assertTrue(page1.getBlock(i).equals(0, 0, page2.getBlock(i), 0, 0, page1.getBlock(0).getSliceLength(0)));
-            }
+        ImmutableList.Builder<Future<String>> futures = ImmutableList.builder();
+        int i = 0;
+        int threadCnt = 10;
+        while (i < threadCnt) {
```
[the remark this answers is not in the document]
Comment (superqtqt) [reply presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java]: ok

File: presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java (hunk 5 above)
Comment (superqtqt) [reply presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java]: i name it test write content,thread %s,%s,that i think it more clear

**Comment by superqtqt** [comment]
thank you for reviewing my code.I am the first time to commit code ,which wasting your time.

File: presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java [hunk 13]
```diff
@@ -51,11 +53,16 @@
 {
     private static final String SERIALIZED_PLAN_FRAGMENT_1 = "test plan fragment 1";
     private static final String SERIALIZED_PLAN_FRAGMENT_2 = "test plan fragment 2";
-    private static final Split SPLIT_1 = new Split(new ConnectorId("test"), new ConnectorTransactionHandle() {}, new TestingSplit(1));
-    private static final Split SPLIT_2 = new Split(new ConnectorId("test"), new ConnectorTransactionHandle() {}, new TestingSplit(2));
+    private static final Split SPLIT_1 = new Split(new ConnectorId("test"), new ConnectorTransactionHandle()
+    {
```
Comment (shixuan-fan) [inline presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java:57]: Maybe revert the changes here and for `SPLIT_2` if they are unintended.

**Review by shixuan-fan — APPROVED** [review]
LGTM. 

No need to apologize, you are the one doing the hard work and helping fixing stuff. We really appreciate that :)

**Comment by shixuan-fan** [comment]
Also, I'm wondering if we should change the commit and PR title to: "Fix multithread writing for fragment result cache"

File: presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java (hunk 13 above)
Comment (superqtqt) [reply presto-main/src/test/java/com/facebook/presto/operator/TestFileFragmentResultCacheManager.java]: fix

**Comment by superqtqt** [comment]
> Also, I'm wondering if we should change the commit and PR title to: "Fix multithread writing for fragment result cache"

ok ,i have changed

**Comment by shixuan-fan** [comment]
Merged. Thanks for the help identifying/fixing bugs :)
